File: server/services/settings-service.ts
import fs from "fs";
import path from "path";
import { DATA_DIR } from "../db.js";

const CONFIG_PATH = path.join(DATA_DIR, "config.json");

const DEFAULTS: Record<string, string> = {
  api_key: process.env["DEEPSEEK_API_KEY"] || "",
  base_url: process.env["DEEPSEEK_BASE_URL"] || "https://api.deepseek.com",
  model: process.env["DEEPSEEK_MODEL"] || "deepseek-v4-flash",
  tavily_api_key: process.env["TAVILY_API_KEY"] || "",
  serper_api_key: process.env["SERPER_API_KEY"] || "",
  brave_search_api_key: process.env["BRAVE_SEARCH_API_KEY"] || "",
  ui_theme: process.env["AKARI_UI_THEME"] || "agent_warm_paper",
  workspace_path: process.env["AKARI_WORKSPACE_PATH"] || "",
};

export interface LlmSettings {
  apiKey: string;
  baseUrl: string;
  model: string;
  tavilyApiKey: string;
  serperApiKey: string;
  braveSearchApiKey: string;
  uiTheme: string;
  workspacePath: string;
}

let cache: LlmSettings | null = null;

export function getSettings(): LlmSettings {
  if (cache) return cache;
  const fileData: Record<string, string> = {};
  if (fs.existsSync(CONFIG_PATH)) {
    try {
      Object.assign(fileData, JSON.parse(fs.readFileSync(CONFIG_PATH, "utf-8")));
    } catch {
      /* ignore corrupt config */
    }
  }
  cache = {
    apiKey: fileData["api_key"] || DEFAULTS["api_key"]!,
    baseUrl: fileData["base_url"] || DEFAULTS["base_url"]!,
    model: fileData["model"] || DEFAULTS["model"]!,
    tavilyApiKey: fileData["tavily_api_key"] || DEFAULTS["tavily_api_key"]!,
    serperApiKey: fileData["serper_api_key"] || DEFAULTS["serper_api_key"]!,
    braveSearchApiKey: fileData["brave_search_api_key"] || DEFAULTS["brave_search_api_key"]!,
    uiTheme: fileData["ui_theme"] || DEFAULTS["ui_theme"]!,
    workspacePath: fileData["workspace_path"] || DEFAULTS["workspace_path"]!,
  };
  return cache;
}

export function saveSettings(s: LlmSettings): void {
  fs.mkdirSync(path.dirname(CONFIG_PATH), { recursive: true });
  fs.writeFileSync(
    CONFIG_PATH,
    JSON.stringify(
      {
        api_key: s.apiKey,
        base_url: s.baseUrl,
        model: s.model,
        tavily_api_key: s.tavilyApiKey,
        serper_api_key: s.serperApiKey,
        brave_search_api_key: s.braveSearchApiKey,
        ui_theme: s.uiTheme,
        workspace_path: s.workspacePath,
      },
      null,
      2,
    ),
  );
  cache = s;
}

export function reloadSettings(): LlmSettings {
  cache = null;
  return getSettings();
}

export function maskKey(key: string): string {
  if (!key) return "";
  if (key.length <= 6) return "***";
  return key.slice(0, 3) + "..." + key.slice(-3);
}

export function resolveApiKey(incoming: string, currentKey: string): string {
  if (!incoming || incoming === maskKey(currentKey)) return currentKey;
  return incoming;
}

File: backend/services/settings_service.py
from __future__ import annotations

import json
import os
from dataclasses import dataclass, asdict

from backend.db.database import DATA_DIR

CONFIG_PATH = DATA_DIR / "config.json"

_DEFAULTS = {
    "api_key": os.getenv("DEEPSEEK_API_KEY", ""),
    "base_url": os.getenv("DEEPSEEK_BASE_URL", "https://api.deepseek.com"),
    "model": os.getenv("DEEPSEEK_MODEL", "deepseek-v4-flash"),
    "tavily_api_key": os.getenv("TAVILY_API_KEY", ""),
    "serper_api_key": os.getenv("SERPER_API_KEY", ""),
    "brave_search_api_key": os.getenv("BRAVE_SEARCH_API_KEY", ""),
    "ui_theme": os.getenv("AKARI_UI_THEME", "agent_warm_paper"),
}


@dataclass
class LlmSettings:
    api_key: str = ""
    base_url: str = "https://api.deepseek.com"
    model: str = "deepseek-v4-flash"
    tavily_api_key: str = ""
    serper_api_key: str = ""
    brave_search_api_key: str = ""
    ui_theme: str = "agent_warm_paper"


_cache: LlmSettings | None = None


def _load_from_file() -> dict:
    if CONFIG_PATH.exists():
        try:
            return json.loads(CONFIG_PATH.read_text())
        except (json.JSONDecodeError, OSError):
            return {}
    return {}


def get_settings() -> LlmSettings:
    global _cache
    if _cache is not None:
        return _cache

    file_data = _load_from_file()
    merged = {**_DEFAULTS, **{k: v for k, v in file_data.items() if v}}
    _cache = LlmSettings(**merged)
    return _cache


def save_settings(s: LlmSettings) -> None:
    global _cache
    CONFIG_PATH.parent.mkdir(parents=True, exist_ok=True)
    CONFIG_PATH.write_text(json.dumps(asdict(s), indent=2, ensure_ascii=False))
    _cache = s


def reload_settings() -> LlmSettings:
    global _cache
    _cache = None
    return get_settings()


def mask_key(key: str) -> str:
    if len(key) <= 6:
        return "***" if key else ""
    return key[:3] + "..." + key[-3:]


def resolve_api_key(incoming: str, current_key: str) -> str:
    """Handle partial update: masked or empty value means unchanged."""
    if not incoming or incoming == mask_key(current_key):
        return current_key
    return incoming
